### ISSUE TITLE
LLoCa 2.0 Wishlist

### ISSUE THREAD

**Issue by t0mnt** [heidelberg-hepml/lloca#14]
Add attention weight dropout support to all backends

Library ParT doesn't use MHA so weights don't use Xavier initialization, minor but a thing (and more importantly trimmer should not run after prepare frames)

see Gtagger-experiments' `lloca_transport_attention`  for complete bit accurate identity frames to ParT, down to init and checkpoints using nn.MultiheadAttention as block's module (EDIT: uses `prepare_frames` and no longer compatible with 2.0)

Copy `for_inference` fix from GTagger-experiments for ParT and ParticleNet

**Comment by t0mnt** [comment]
Draft, second wishlist
Identity frames already turn global on, computing `q·k` instead of `q·ηk` so nothing to note there
Also pop backend for get_attention_backend(**attn_kwargs) forwarding 
this is already done in lgatr dev and its something that came to mind

**Comment by t0mnt** [comment]
See https://github.com/heidelberg-hepml/lloca-experiments/pull/95 for `for_inference` tweak

**Comment by t0mnt** [comment]
And maybe a guide for implementing tensorial message passing beyond git diffs

**Comment by t0mnt** [comment]
`lloca.equivectors.lgatr` imports `_load_inner_product_factors` from lgatr's private API

SO13 is more invariant than PDFrames in float64 by my measurement despite the latter being recommended ( ~5e-3 against ~1e-6 for learnedso13). Just out of curiosity so I can learn, why is this?

add a `frames.dim()` check to `prepare_frames()`

**Comment by t0mnt** [comment]
NaN issue :(
`restframe_boost` bounds γ in the forward pass but leaves its derivative at the maximum, so a nearly-lightlike predicted boost produces a finite loss and an inf/NaN gradient. Enabling `gamma_max` helps but produces NaN errors of its own

`restframe_boost` in `polar_decomposition.py`  clamp caps γ at 1e5 by pinning where `rsqrt` is steepest, the derivative there is about -5e14, and scale differentiates to ~1e25, even larger. fp32 tops out at 3.4e38, so two of those in one product overflow & NaN
```

one_minus_beta2 = torch.clamp_min(1 - beta2, min=1e-10)
gamma = torch.rsqrt(one_minus_beta2)
...
scale = (gamma - 1) / torch.clamp_min(beta2, min=1e-10)

```
gamma_max helps by an order of magnitude but still NaNs, `clamp_boost` in `equi_frames.py` since `sqrt` has infinite derivative at 0 (doesn't return `x` before `mass `is used when `gamma_max !=None`)
```

mass = lorentz_squarednorm(x).clamp(min=0).sqrt().unsqueeze(-1)
...
x_reg = mass * torch.cat((gamma_reg, gamma_reg * beta_reg), dim=-1)

```
fortunately only on 1/4 of my models where LLoCa can potentially help but still something I found worth notifying
